provider/backup_utils.py

```diff
@@ -406,3 +409,52 @@ def format_storage_volume(img, filesystem, partition="mbr"):
             partition="mbr")
     finally:
         process.system("setenforce %s" % selinux_mode, shell=True)
+
+
+def copyif(params, nbd_image, target_image, bitmap=None):
+    """
+    Python implementation of copyif3.sh
+    :params params: utils_params.Params object
+    :params nbd_image: nbd image tag
+    :params target_image: target image tag
+    :params bitmap: bitmap name
+    """
+    qemu_img = utils_misc.get_qemu_img_binary(params)
+    qemu_io = utils_misc.get_qemu_io_binary(params)
+    img_obj = qemu_storage.QemuImg(params.object_params(target_image),
+                                   data_dir.get_data_dir(), target_image)
+    nbd_img_obj = qemu_storage.QemuImg(params.object_params(nbd_image),
+                                       None, nbd_image)
+
+    if bitmap is None:
+        args = '-f %s %s' % (nbd_img_obj.image_format,
+                             nbd_img_obj.image_filename)
+        state = True
+    else:
+        opts = qemu_storage.filename_to_file_opts(
+            nbd_img_obj.image_filename)
+        opt = params.get('dirty_bitmap_opt', 'x-dirty-bitmap')
+        opts[opt] = 'qemu:dirty-bitmap:%s' % bitmap
+        args = "'json:%s'" % json.dumps(opts)
+        state = False
+
+    img_obj.base_image_filename = nbd_img_obj.image_filename
+    img_obj.base_format = nbd_img_obj.image_format
+    img_obj.base_tag = nbd_img_obj.tag
+    img_obj.rebase(img_obj.params)
+
+    map_cmd = '{qemu_img} map --output=json {args}'.format(
+        qemu_img=qemu_img, args=args)
+    result = process.run(map_cmd, ignore_status=False, shell=True)
+
+    for item in json.loads(result.stdout.decode().strip()):
+        if item['data'] is not state:
+            continue
+        io_cmd = '{io} -C -c "r {s} {l}" -f {fmt} {fn}'.format(
+            io=qemu_io, s=item['start'], l=item['length'],
```

Comment:
@zhencliu Zhenchao, according to https://bugzilla.redhat.com/show_bug.cgi?id=1657095#c10, we need to divide the qemu-io length to be within 0 ~ 2147483136. So will you modify it? Or we can merge the patch first and then fix it together with the copyif() in other cases in a future patch.

Reply:
Nini, let me enhance it in this patch, will let you know once I finish it. Thanks.

Reply:
change done.